src/appleseed/foundation/image/colormap.cpp

```diff
@@ -0,0 +1,165 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2019 Achal Pandey, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "colormap.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/colorspace.h"
+#include "foundation/math/scalar.h"
+
+// Standard headers.
+#include <cassert>
+#include <limits>
+
+namespace foundation
+{
+
+void ColorMap::find_min_max_red_channel(
+    Image&          image,
+    const AABB2u&   crop_window,
+    float&          min_val,
+    float&          max_val)
+{
+    for_each_pixel(image, crop_window, [&min_val, &max_val](const Color4f& val)
+    {
+        min_val = min(val[0], min_val);
+        max_val = max(val[0], max_val);
+    });
+}
+
+void ColorMap::find_min_max_relative_luminance(
+    Image&              image,
+    const AABB2u&       crop_window,
+    float&              min_luminance,
+    float&              max_luminance)
+{
+    min_luminance = luminance(Color4f(+std::numeric_limits<float>::max()).rgb());
```

Comment:
It would make more sense to directly initialize `min_luminance` and `max_luminance` to +/- `std::numeric_limits<float>::max()` rather than go through the `luminance()` function.